**SQL commands in pandas**. 

SQL emulator using Python pandas functions. Has the advantages:

• The functions can easily be stored for future use
   
• Can pass the output of one fucntion in input to another

• Sigificantly less code 
   
• Functions can be made interactive 

More generally:

• Can (easily) run on command line and within analysis script for pipelines

• The complexity and size of the SQL standard means that most implementers do not support the entire standard – so may be some incomptability with peers



See https://github.com/steviecurran/sql2csv to convert .sql to .csv, via the C shell

<a id='cont'></a>
**Function contents**

| SELECT		          | Aggregate              | ORDER and GROUP       | JOIN                       | Others| 
| :---                 | :---                   | :--- 	           | :---                       | :-- | 
|**[WHERE](#WHERE)**  |**[COUNT](#COUNT)**  |**[ORDER BY](#OB)**    |**[INNER](#INNER)**  |**[DISTINCT](#DIST)**| 
|**[OR](#OR)**  |**[SUM](#AGG)**         |**[GROUP COUNT](#GC)** |**[OUTER](#OUTER)**         |**[FILTER](#FIL)**|
|**[AND and OR](#AO)** |**[MIN and MAX](#AGG)** |**[GROUP AGG](#GA)** |**[LEFT/RIGHT](#LR)**  |**[CASE](#CAS)**|  
| **[IN/NOT IN](#IN)** |**[AVG](#AGG)** |**[HAVING](#HA)**|**[MULTIPLE tables](#MUL)** |**[Window functions](#WF)** |
| **[LIKE/NOT LIKE](#LIKE)**      |**[MEDIAN](#AGG)**      |**[DUPLICATES](#DUP)** |**[JOIN WHERE](#JW)**       | |
| **[BETWEEN/NOT BETWEEN](#BET)** |                        |                  	   | **[CROSS JOIN](#CJ)**      | |
| **[IS NULL/NOT NULL](#NULL)**   |                        |                       | **[Aggregate and JOIN](#AJ)**|
|

In [2]:
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None

Load some example data

In [3]:
emp = pd.read_csv("employees.csv");
sal = pd.read_csv("salaries.csv");
dep = pd.read_csv("departments.csv")
man = pd.read_csv("dept_manager.csv")
dept_emp = pd.read_csv("dept_emp.csv")
titles = pd.read_csv("titles.csv")

# SELECT 

**SELECT FROM**

In [4]:
def SELECT(data,field,entry):
    return data[data[field] == entry]

In [1086]:
SELECT(emp,'first_name',"Elvis") # E.g. select from employee table anyone with first name Elvis

emp_no  birth_date first_name last_name gender   hire_date
29       10030  1958-07-14      Elvis   Demeyer      M  1994-02-17
1049     11050  1952-09-23      Elvis   Katiyar      M  1986-05-06
2869     12870  1960-03-24      Elvis      Pfau      F  1990-04-09
3619     13620  1956-10-31      Elvis  Dolinsky      M  1985-06-12
4433     14434  1954-07-08      Elvis  Maginnis      M  1987-11-27
...        ...         ...        ...       ...    ...         ...
295430  495406  1956-05-04      Elvis   Gerlach      M  1990-09-17
296527  496503  1956-09-01      Elvis  Verhoeff      M  1986-10-27
297934  497910  1961-05-12      Elvis   Cannard      F  1990-09-23
298645  498621  1955-09-07      Elvis      Zykh      M  1993-08-31
299115  499091  1952-09-24      Elvis   Kohling      F  1986-02-25

[246 rows x 6 columns]

 [Back to contents](#cont)
 
 **<a id='WHERE'>SELECT WHERE</a>**

In [1146]:
def SELECT_WHERE(data,field1,entry1,field2,entry2): 
    return data[(data[field1] == entry1) & (data[field2] == entry2)]

In [9]:
SELECT_WHERE(emp,'first_name',"Denis",'gender',"M") # E.g. from employee table first name Denis and gender M

emp_no  birth_date first_name     last_name gender   hire_date
5082     15083  1958-11-24      Denis      Nicolson      M  1994-03-02
7223     17224  1955-04-06      Denis          Back      M  1985-07-29
8120     18121  1953-08-26      Denis       Stanfel      M  1990-05-21
8186     18187  1965-01-08      Denis       Falster      M  1989-10-12
9797     19798  1962-12-22      Denis      Schmezko      M  1986-10-30
...        ...         ...        ...           ...    ...         ...
290517  490493  1956-07-06      Denis        Auyong      M  1989-04-15
290863  490839  1958-06-19      Denis      Narwekar      M  1987-12-09
291119  491095  1962-05-30      Denis        Kuszyk      M  1992-12-05
296968  496944  1954-09-02      Denis  Fortenbacher      M  1986-02-09
297316  497292  1964-05-28      Denis     Gustavson      M  1990-07-25

[140 rows x 6 columns]

In [39]:
SELECT_WHERE(emp,'first_name',"Elvis",'last_name',"Presley") # Sorry, just had to

Empty DataFrame
Columns: [emp_no, birth_date, first_name, last_name, gender, hire_date]
Index: []

[Back to contents](#cont)

**WHERE with comparison operator**

In [92]:
def FILTER(data,field,entry,action):
    act = "print(data[data[field] %s entry])" %(action)
    exec(act)

In [93]:
FILTER(emp,'hire_date','2000-01-01','>') # E.g. employees hired after 2000-01-01 exclusive

        emp_no  birth_date first_name   last_name gender   hire_date
37290    47291  1960-09-09        Ulf      Flexer      M  2000-01-12
50133    60134  1964-04-21      Seshu    Rathonyi      F  2000-01-02
62328    72329  1953-02-09      Randi        Luit      F  2000-01-02
105072  205048  1960-09-12      Ennio      Alblas      F  2000-01-06
122989  222965  1959-08-07    Volkmar       Perko      F  2000-01-13
126657  226633  1958-06-10     Xuejun  Benzmuller      F  2000-01-04
127568  227544  1954-11-17     Shahab     Demeyer      M  2000-01-08
223014  422990  1953-04-09      Jaana    Verspoor      F  2000-01-11
224469  424445  1953-04-27      Jeong     Boreale      M  2000-01-03
228401  428377  1957-05-09      Yucai     Gerlach      M  2000-01-23
263831  463807  1964-06-12     Bikash      Covnot      M  2000-01-28
299577  499553  1954-05-06   Hideyuki   Delgrande      F  2000-01-22


In [94]:
FILTER(emp,'hire_date','2000-01-01','>=') # E.g. employees hired after 2000-01-01 inclusive

        emp_no  birth_date   first_name   last_name gender   hire_date
37290    47291  1960-09-09          Ulf      Flexer      M  2000-01-12
50133    60134  1964-04-21        Seshu    Rathonyi      F  2000-01-02
62328    72329  1953-02-09        Randi        Luit      F  2000-01-02
98200   108201  1955-04-14  Mariangiola     Boreale      M  2000-01-01
105072  205048  1960-09-12        Ennio      Alblas      F  2000-01-06
122989  222965  1959-08-07      Volkmar       Perko      F  2000-01-13
126657  226633  1958-06-10       Xuejun  Benzmuller      F  2000-01-04
127568  227544  1954-11-17       Shahab     Demeyer      M  2000-01-08
223014  422990  1953-04-09        Jaana    Verspoor      F  2000-01-11
224469  424445  1953-04-27        Jeong     Boreale      M  2000-01-03
228401  428377  1957-05-09        Yucai     Gerlach      M  2000-01-23
263831  463807  1964-06-12       Bikash      Covnot      M  2000-01-28
299577  499553  1954-05-06     Hideyuki   Delgrande      F  2000-01-22


[Back to contents](#cont)

**WHERE with comparison operator and condition**

In [95]:
def FILTER_COND(data,field1,entry1,field2,entry2,action): # FILTER with condition
    tmp = data[data[field1] == entry1]; 
    act = "print(tmp[tmp[field2] %s entry2])" %(action)
    exec(act)

In [96]:
FILTER_COND(emp,'gender','F','hire_date','2000','<') 
# E.g. Female employees who were hired before the year 2000

        emp_no  birth_date first_name  last_name gender   hire_date
1        10002  1964-06-02    Bezalel     Simmel      F  1985-11-21
5        10006  1953-04-20     Anneke    Preusig      F  1989-06-02
6        10007  1957-05-23    Tzvetan  Zielinski      F  1989-02-10
8        10009  1952-04-19     Sumant       Peac      F  1985-02-18
9        10010  1963-06-01  Duangkaew   Piveteau      F  1989-08-24
...        ...         ...        ...        ...    ...         ...
300012  499988  1962-09-28   Bangqing    Kleiser      F  1986-06-06
300015  499991  1962-02-26      Pohua    Sichman      F  1989-01-12
300016  499992  1960-10-12     Siamak   Salverda      F  1987-05-10
300018  499994  1952-02-26      Navin    Argence      F  1990-04-24
300019  499995  1958-09-24     Dekang   Lichtner      F  1993-01-12

[120044 rows x 6 columns]


[Back to contents](#cont)

**<a id='OR'>SELECT OR</a>** 

In [40]:
def SELECT_OR(data,field,entry1,entry2):
    return data[(data[field] == entry1) | (data[field] == entry2)]

In [41]:
SELECT_OR(emp,'first_name',"Denis","Elvis") # E.g. from employee table first name Denis or Elvis

emp_no  birth_date first_name  last_name gender   hire_date
29       10030  1958-07-14      Elvis    Demeyer      M  1994-02-17
1049     11050  1952-09-23      Elvis    Katiyar      M  1986-05-06
1687     11688  1958-09-04      Denis   Coullard      F  1994-10-29
2869     12870  1960-03-24      Elvis       Pfau      F  1990-04-09
3619     13620  1956-10-31      Elvis   Dolinsky      M  1985-06-12
...        ...         ...        ...        ...    ...         ...
297316  497292  1964-05-28      Denis  Gustavson      M  1990-07-25
297934  497910  1961-05-12      Elvis    Cannard      F  1990-09-23
298284  498260  1959-04-15      Denis    Fetvedt      F  1988-05-04
298645  498621  1955-09-07      Elvis       Zykh      M  1993-08-31
299115  499091  1952-09-24      Elvis    Kohling      F  1986-02-25

[478 rows x 6 columns]

[Back to contents](#cont)

**<a id='AO'>SELECT AND and OR</a>** 

In [42]:
def SELECT_AND_OR(data,field1,entry1,field2,entry2,entry3): 
    return data[(data[field1] == entry1) & ((data[field2] == entry2) | (data[field2] == entry3))]

In [43]:
SELECT_AND_OR(emp,'gender',"F",'first_name',"Kellie","Aruna") # E.g. gender F with first name Kellie or Aruna

emp_no  birth_date first_name  last_name gender   hire_date
224      10225  1957-02-13     Kellie     Chinen      F  1986-06-19
788      10789  1964-05-19      Aruna    Journel      F  1987-02-02
2037     12038  1956-04-13      Aruna   Businaro      F  1987-01-09
4079     14080  1957-02-24      Aruna     Motley      F  1987-03-09
4458     14459  1964-09-21      Aruna    Boreale      F  1986-12-06
...        ...         ...        ...        ...    ...         ...
287044  487020  1960-08-12     Kellie    Busillo      F  1985-05-09
288205  488181  1960-05-02      Aruna  Delgrande      F  1986-12-15
292705  492681  1957-04-06     Kellie    Koshino      F  1991-03-28
294769  494745  1962-09-16      Aruna   Massonet      F  1993-09-10
294891  494867  1953-08-24     Kellie    Chenney      F  1987-06-28

[167 rows x 6 columns]

[Back to contents](#cont)

**<a id='IN'>SELECT IN/NOT IN</a>**

In [97]:
def SELECT_IN(data,field,entry):
    return data[data[field].isin(entry)]

In [45]:
SELECT_IN(emp,'first_name', ['Cathie','Mark','Nathan']) # E.g. first name Cathie, Mark or Nathan in the emplyee table

emp_no  birth_date first_name  last_name gender   hire_date
134      10135  1956-12-23     Nathan  Monkewich      M  1988-02-19
414      10415  1957-11-12       Mark      Coorg      M  1993-10-25
427      10428  1957-06-25     Cathie      Brlek      M  1992-06-04
1094     11095  1961-12-31     Nathan    Flowers      M  1994-05-28
1502     11503  1953-08-01     Nathan      Picht      F  1993-01-25
...        ...         ...        ...        ...    ...         ...
298225  498201  1952-03-22       Mark     Schahn      F  1990-12-24
298880  498856  1952-07-26     Cathie    Wiegley      M  1999-03-09
298941  498917  1955-12-28     Nathan   Papadias      F  1987-01-09
299858  499834  1956-04-04     Nathan     Chimia      F  1995-05-12
300010  499986  1952-07-22     Nathan      Ranta      F  1985-08-11

[691 rows x 6 columns]

In [49]:
def SELECT_NOT_IN(data,field,entry):
    return data[~data[field].isin(entry)]

In [51]:
SELECT_NOT_IN(emp,'first_name', ['Cathie','Mark','Nathan']) # E.g. all first names which are not Cathie, Mark or Nathan

emp_no  birth_date first_name last_name gender    hire_date
0        10001  1953-09-02     Georgi   Facello      M   1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F   1985-11-21
2        10003  1959-12-03      Parto   Bamford      M   1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M   1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M   1989-09-12
...        ...         ...        ...       ...    ...          ...
300019  499995  1958-09-24     Dekang  Lichtner      F   1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M   1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M   1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M   1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30;

[299333 rows x 6 columns]

[Back to contents](#cont)

**<a id='LIKE'>SELECT LIKE/NOT LIKE</a>**

In [59]:
def SELECT_LIKE(data,field,entry,action):
    act = "print(data[data[field].str.%s(entry)])" %(action)
    exec(act)

In [60]:
SELECT_LIKE(emp,'first_name', 'Mark','startswith') # E.g. First name starts with 'Mark'

        emp_no  birth_date first_name    last_name gender   hire_date
231      10232  1956-03-11      Marko        Auria      F  1992-06-04
414      10415  1957-11-12       Mark        Coorg      M  1993-10-25
521      10522  1955-07-24      Marke     Cesareni      F  1986-12-12
1886     11887  1957-05-25      Marke    Kragelund      M  1994-10-17
2006     12007  1952-06-18      Marko     Dehkordi      M  1989-05-17
...        ...         ...        ...          ...    ...         ...
298713  498689  1954-10-23      Marko     Hatcliff      F  1985-06-03
298957  498933  1952-05-24      Marke  Grospietsch      F  1996-06-02
299004  498980  1957-05-29      Marko     Eppinger      M  1988-07-13
299475  499451  1953-04-28      Marko   Goldhammer      M  1986-10-19
299644  499620  1961-07-30      Marke         Ibel      F  1996-01-30

[690 rows x 6 columns]


In [61]:
SELECT_LIKE(emp,'first_name','ar','contains') # E.g. First name contains 'ar'

        emp_no  birth_date  first_name last_name gender   hire_date
2        10003  1959-12-03       Parto   Bamford      M  1986-08-28
10       10011  1953-11-07        Mary     Sluis      F  1990-01-22
12       10013  1963-06-07   Eberhardt    Terkki      M  1985-10-20
28       10029  1956-12-13       Otmar    Herbst      M  1985-11-20
30       10031  1959-01-27     Karsten    Joslin      M  1991-09-01
...        ...         ...         ...       ...    ...         ...
299970  499946  1956-08-10      Barton   Soicher      F  1986-02-21
299985  499961  1962-10-02     Holgard    Nanard      F  1988-07-04
299992  499968  1959-03-07  Dharmaraja      Ertl      M  1991-10-08
300001  499977  1956-06-05     Martial   Weisert      F  1996-09-17
300021  499997  1961-08-03     Berhard    Lenart      M  1986-04-21

[25828 rows x 6 columns]


In [63]:
SELECT_LIKE(emp,'first_name','ar','endswith') # E.g. First name ends with 'ar'

        emp_no  birth_date  first_name  last_name gender   hire_date
28       10029  1956-12-13       Otmar     Herbst      M  1985-11-20
73       10074  1955-08-28     Mokhtar  Bernatsky      F  1990-08-13
174      10175  1960-01-11  Aleksandar  Ananiadou      F  1988-01-11
265      10266  1958-02-24     Sukumar    Rassart      M  1990-05-25
401      10402  1953-07-23     Volkmar    Ebeling      M  1987-01-02
...        ...         ...         ...        ...    ...         ...
299676  499652  1963-01-16      Oddvar    Ushiama      F  1993-11-27
299786  499762  1952-08-24   Subhankar    Munawer      M  1985-07-06
299830  499806  1958-10-11      Ingmar  Barbanera      F  1995-04-14
299908  499884  1961-03-15     Adhemar       Gist      F  1988-11-19
299951  499927  1953-03-18     Manohar  Heemskerk      M  1988-03-13

[5320 rows x 6 columns]


In [65]:
def SELECT_NOT_LIKE(data,field,entry,action):
    act = "print(data[~data[field].str.%s(entry)])" %(action)
    exec(act)

[Back to contents](#cont)

**<a id='BET'>SELECT BETWEEN/NOT BETWEEN</a>** 

In [51]:
def SELECT_BETWEEN(data,field,entry1,entry2,inc):
    if inc == "y": # INCLUSIVE
        return data[(data[field] >= entry1) & (data[field] <= entry2)]
    else: # EXCLUSIVE
        return data[(data[field] > entry1) & (data[field] < entry2)]

In [52]:
SELECT_BETWEEN(emp,'hire_date','1990-01-01','2000-01-01','y') 
# E.g. All hire dates between 1990-01-01 and 2000-01-01 inclusive

emp_no  birth_date first_name    last_name gender    hire_date
7        10008  1958-02-19     Saniya     Kalloufi      M   1994-09-15
10       10011  1953-11-07       Mary        Sluis      F   1990-01-22
11       10012  1960-10-04   Patricio    Bridgland      M   1992-12-18
15       10016  1961-05-02   Kazuhito  Cappelletti      M   1995-01-27
16       10017  1958-07-06  Cristinel    Bouloucos      F   1993-08-03
...        ...         ...        ...          ...    ...          ...
300018  499994  1952-02-26      Navin      Argence      F   1990-04-24
300019  499995  1958-09-24     Dekang     Lichtner      F   1993-01-12
300020  499996  1953-03-07       Zito         Baaz      M   1990-09-27
300022  499998  1956-09-05   Patricia      Breugel      M   1993-10-13
300023  499999  1958-05-01     Sachin      Tsukuda      M  1997-11-30;

[135215 rows x 6 columns]

[Back to contents](#cont)

**E.g. assign emplpyee number 110022 as a manger to all employees from 11001 to 11020 and employee number 110039 to all employees from 11021 to 11040** 

Takes a lot of SQL!

In [455]:
empA = SELECT_BETWEEN(emp,'emp_no',11001,11020,'y'); empA['manager_ID'] = 110022
empB = SELECT_BETWEEN(emp,'emp_no',11021,11040,'y'); empB['manager_ID'] = 110039
print(pd.concat([empA, empB], ignore_index=True))

    emp_no  birth_date   first_name     last_name gender   hire_date  \
0    11001  1956-04-16      Baziley       Buchter      F  1987-02-23   
1    11002  1952-02-26        Bluma      Ulupinar      M  1996-12-23   
2    11003  1960-11-13  Mariangiola         Gulla      M  1987-05-24   
3    11004  1954-08-05       JoAnna       Decleir      F  1992-01-19   
4    11005  1958-03-12        Byong       Douceur      F  1986-07-27   
5    11006  1962-12-26   Christoper   Butterworth      F  1989-08-02   
6    11007  1962-03-16      Olivera     Maccarone      M  1991-04-11   
7    11008  1962-07-11      Gennady      Menhoudj      M  1988-09-18   
8    11009  1954-08-30        Alper      Axelband      F  1986-09-09   
9    11010  1963-03-20       Jaques      Narwekar      F  1986-12-18   
10   11011  1956-06-13         Moti    Hofstetter      M  1989-11-06   
11   11012  1953-09-12      Taegyun     Speckmann      F  1986-12-28   
12   11013  1961-03-31     Breannda         Vesel      M  1996-0

/var/folders/pb/dhr60dvd19370mv4qdxrn34m0000gn/T/ipykernel_67521/1977768428.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  empA = SELECT_BETWEEN(emp,'emp_no',11001,11020,'y'); empA['manager_ID'] = 110022
/var/folders/pb/dhr60dvd19370mv4qdxrn34m0000gn/T/ipykernel_67521/1977768428.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  empB = SELECT_BETWEEN(emp,'emp_no',11021,11040,'y'); empB['manager_ID'] = 110039


In [113]:
def SELECT_NOT_BETWEEN(data,field,entry1,entry2,inc):
    if inc == "y":
        return data[(data[field] <= entry1) | (data[field] >= entry2)]
    else: 
        return data[(data[field] < entry1) | (data[field] > entry2)]    

In [76]:
SELECT_NOT_BETWEEN(emp,'hire_date','1990-01-01','2000-01-01','n')
# E.g. All hire dates note between 1990-01-01 and 2000-01-01 exclusive

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300012  499988  1962-09-28   Bangqing   Kleiser      F  1986-06-06
300014  499990  1963-11-03     Khaled   Kohling      M  1985-10-10
300015  499991  1962-02-26      Pohua   Sichman      F  1989-01-12
300016  499992  1960-10-12     Siamak  Salverda      F  1987-05-10
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21

[164809 rows x 6 columns]

[Back to contents](#cont)

**<a id='NULL'>SELECT IS NULL/NOT NULL</a>**

In [78]:
def SELECT_NULL(data,field):
    return data[data[field] != data[field]]

In [80]:
SELECT_NULL(emp,'first_name') # E.g. extract rows where the first name is null

Empty DataFrame
Columns: [emp_no, birth_date, first_name, last_name, gender, hire_date]
Index: []

In [81]:
def SELECT_NOT_NULL(data,field):
    return data[data[field] == data[field]]

In [82]:
SELECT_NOT_NULL(emp,'first_name')

emp_no  birth_date first_name last_name gender    hire_date
0        10001  1953-09-02     Georgi   Facello      M   1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F   1985-11-21
2        10003  1959-12-03      Parto   Bamford      M   1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M   1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M   1989-09-12
...        ...         ...        ...       ...    ...          ...
300019  499995  1958-09-24     Dekang  Lichtner      F   1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M   1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M   1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M   1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30;

[300024 rows x 6 columns]

[Back to contents](#cont)

**<a id='DIST'>DISTINCT</a>** 

**Show all uniques entries in a field**

In [13]:
def SELECT_DISTINCT(data,field): 
    return data[field].unique()

In [91]:
SELECT_DISTINCT(emp,'gender')

array(['M', 'F'], dtype=object)

**<a id='FIL'>FILTER</a>**

In [8]:
def FILTER(data,field,entry,action):
    if action == ">":
        tmp = data[data[field] > entry]
    elif action == ">=":
        tmp = data[data[field] >= entry]
    elif action == "<":
        tmp = data[data[field] < entry]
    elif action == "<=":
        tmp = data[data[field] <= entry]
    else:
        tmp = data[data[field] == entry]   
    return tmp

In [9]:
FILTER(emp,'hire_date','2000-01-01','>') # Just a general filter, e,g employess who started after 2000-01-01

emp_no  birth_date first_name   last_name gender   hire_date
37290    47291  1960-09-09        Ulf      Flexer      M  2000-01-12
50133    60134  1964-04-21      Seshu    Rathonyi      F  2000-01-02
62328    72329  1953-02-09      Randi        Luit      F  2000-01-02
105072  205048  1960-09-12      Ennio      Alblas      F  2000-01-06
122989  222965  1959-08-07    Volkmar       Perko      F  2000-01-13
126657  226633  1958-06-10     Xuejun  Benzmuller      F  2000-01-04
127568  227544  1954-11-17     Shahab     Demeyer      M  2000-01-08
223014  422990  1953-04-09      Jaana    Verspoor      F  2000-01-11
224469  424445  1953-04-27      Jeong     Boreale      M  2000-01-03
228401  428377  1957-05-09      Yucai     Gerlach      M  2000-01-23
263831  463807  1964-06-12     Bikash      Covnot      M  2000-01-28
299577  499553  1954-05-06   Hideyuki   Delgrande      F  2000-01-22

[Back to contents](#cont) 

# AGGREGATE

**<a id='COUNT'>COUNT</a>**

In [102]:
def COUNT(data,field,unique):
    if unique == "y":
        return len(data[field].unique())
    else:    
        return data[field].count()

In [103]:
COUNT(emp,'first_name','y') # E.g. count the number of unique first names

1275

In [165]:
COUNT(emp,'first_name','n') # E.g. count the number of first names

300024

**COUNT with filter**

In [225]:
def FILTER_COUNT(data,field,entry,action): 
    act = "print(data[data[field] %s entry].count())" %(action)
    exec(act)

In [227]:
FILTER_COUNT(sal,'salary',150000,'>') # E.g. count no of salaries greater than 150,000

emp_no       16
salary       16
from_date    16
to_date      16
dtype: int64


[Back to contents](#cont)

**<a id='AGG'>SUM, AVG, MEDIAN, MIN or MAX</a>**

In [20]:
def AGG(data,field,action):
    #act = "print(data[field].%s(()))" %(action); exec(act) # NOT WORKING IN NOTEBOOK
    if action == "sum":
        return data[field].sum()
    elif action == "ave":
        return data[field].mean()
    elif action == "median":
        return data[field].median()
    elif action == "min":
        return data[field].min()
    elif action == "max":
        return data[field].max() 

In [21]:
AGG(sal,'salary','ave') # E.g. the average salary from the 'sal' table

63761.2043294429

In [465]:
AGG(sal,'salary','median') # E.g. the median salary from the 'sal' table

61070.0

In [400]:
# E.g. #What is the average annual salary paid to employees who started after the 1st of January 2000?
tmp = sal[sal['from_date'] > '2000-01-01']
df = AGG(tmp,'salary','ave'); 
print(df)

70239.19785541651


# ORDER and GROUP

[Back to contents](#cont)
 
**<a id='OB'>ORDER BY</a>**

In [257]:
def ORDER_BY(data,field,ascending):
    return data.sort_values(by=field, ascending=ascending)

In [258]:
ORDER_BY(emp,'first_name',True) # E.g. order employee table in by primary field in ascending order  

emp_no  birth_date first_name  last_name gender   hire_date
290185  490161  1963-01-18      Aamer     Rubsam      F  1996-12-08
84466    94467  1963-04-24      Aamer   Benantar      F  1987-01-28
93640   103641  1955-11-11      Aamer    Loncour      M  1987-06-14
230575  430551  1960-10-22      Aamer      Krogh      M  1987-07-29
120795  220771  1956-12-17      Aamer   Engberts      M  1998-05-14
...        ...         ...        ...        ...    ...         ...
275773  475749  1963-01-09     Zvonko       Bahr      M  1994-03-05
96764   106765  1962-05-17     Zvonko   Salvesen      F  1987-12-11
143716  243692  1961-04-19     Zvonko  Antonisse      F  1994-05-22
132055  232031  1952-03-31     Zvonko      Alpay      F  1987-02-21
30998    40999  1963-08-12     Zvonko     Cmelik      M  1986-11-13

[300024 rows x 6 columns]

In [262]:
ORDER_BY(emp, ['first_name','last_name'],True) 
# E.g. order employee table in by primary and secondary fields in ascending order 

emp_no  birth_date first_name   last_name gender   hire_date
59255    69256  1962-04-14      Aamer       Anger      M  1998-03-16
286608  486584  1952-08-12      Aamer      Armand      M  1990-09-15
137189  237165  1962-02-23      Aamer     Azevdeo      F  1991-06-28
213712  413688  1955-06-26      Aamer       Azuma      M  1989-12-10
181387  281363  1956-05-18      Aamer        Baak      F  1994-03-10
...        ...         ...        ...         ...    ...         ...
270706  470682  1960-03-24     Zvonko   Yoshizawa      M  1987-03-12
130730  230706  1961-07-06     Zvonko       Yurov      M  1989-09-08
99488   109489  1962-11-09     Zvonko  Zambonelli      M  1987-04-11
182120  282096  1953-04-13     Zvonko       Zobel      M  1987-04-13
147573  247549  1960-01-07     Zvonko     Zuberek      F  1996-06-13

[300024 rows x 6 columns]

**<a id='GC'>GROUP COUNT</a>**

In [37]:
def GROUP_COUNT(data,field):
    return data.groupby(field).agg(Count=(field,'count')).reset_index()

In [271]:
GROUP_COUNT(emp,'first_name') # E.g. count the employees with the same first names

first_name  Count
0          Aamer    228
1          Aamod    216
2      Abdelaziz    227
3     Abdelghani    247
4     Abdelkader    222
...          ...    ...
1270       Ziyad    229
1271     Zongyan    254
1272      Zorica    225
1273       Zsolt    236
1274      Zvonko    258

[1275 rows x 2 columns]

In [389]:
# E.g. Select the employee numbers of all individuals who have signed more than 1 contract after the 1st of January 2000.
tmp = GROUP_COUNT(dept_emp,'emp_no')
tmp2 = tmp[tmp['Count'] >1];
print(tmp2)

        emp_no  Count
9        10010      2
17       10018      2
28       10029      2
39       10040      2
49       10050      2
...        ...    ...
299979  499955      2
299988  499964      2
299999  499975      2
300007  499983      2
300016  499992      2

[31579 rows x 2 columns]


[Back to contents](#cont)   

**<a id='GA'>GROUP AGG</a>**

**Aggregate functions on groups**

In [53]:
def GROUP_AGG(data,field,sum_field,action):
    astring = 'data.groupby(field).%s(numeric_only=True)' %(action)
    return eval(astring)

In [7]:
GROUP_AGG(sal,'emp_no','salary','mean') # E.g. the mean salary by employee number

salary
emp_no              
10001   75388.941176
10002   68854.500000
10003   43030.285714
10004   56512.250000
10005   87275.769231
...              ...
201767  47097.000000
201768  39818.000000
201769  54001.500000
201770  75524.100000
201771  54893.153846

[101796 rows x 1 columns]

In [470]:
GROUP_AGG(sal,'emp_no','salary','max') # E.g. the maximum salary by employee number

salary
emp_no        
10001    88958
10002    72527
10003    43699
10004    74057
10005    94692
...        ...
201767   47097
201768   40000
201769   62287
201770   85571
201771   66792

[101796 rows x 1 columns]

In [474]:
tmp = GROUP_AGG(sal,'emp_no','salary','mean') # E.g. employees show average salary exceeds 120,000
FILTER(tmp,'salary',120000,'>') 

salary
emp_no               
11486   124462.111111
12149   123656.928571
13555   120264.214286
14487   125684.875000
14508   120655.111111
...               ...
106201  125139.000000
107140  124333.666667
107430  120795.666667
109334  141835.333333
200625  123850.363636

[101 rows x 1 columns]

**<a id='HA'>HAVING</a>**

In [14]:
def HAVING (data,field,action,value):
      grouped = data.groupby(field).agg(Count=(field,'count'))
      act = "print(grouped[grouped['Count'] %s value])" %(action)
      exec(act)

In [15]:
HAVING(emp,'first_name','>',250) # E.g. Employees in which over 250 have the same first name

            Count
first_name       
Adam          251
Akemi         259
Anyuan        278
Arie          255
Arno          261
...           ...
Yurij         266
Zhenhua       257
Zito          251
Zongyan       254
Zvonko        258

[193 rows x 1 columns]


In [16]:
def SAME_COUNT(data,field,action,value):
    act = "print(data[data[field] %s value].groupby(field).agg(same=(field,'count')).reset_index())" %(action)
    exec(act)

In [17]:
SAME_COUNT(sal,'salary', '>=', 80000) 
# E.g. count the number of employees with the same salary where this is greater or euqal to 80,0000

       salary  same
0       80000    14
1       80001    12
2       80002    10
3       80003    10
4       80004    13
...       ...   ...
36613  155190     1
36614  155377     1
36615  155709     1
36616  157821     1
36617  158220     1

[36618 rows x 2 columns]


[Back to contents](#cont) 

**<a id='DUP'>DUPLICATES</a>**

In [159]:
def DROP_DUPS(table): # Will drop duplicate rows keeping the first
    table['duplicated'] = table.duplicated(); print(table)
    del table['duplicated']
    return table.drop_duplicates()

In [160]:
# Testing - copy first 10 rows of emp table and copy first row to the end
emp_dup = emp.head(10); emp_dup.loc[len(emp_dup)] = emp_dup.loc[0];
DROP_DUPS(emp_dup)

    emp_no  birth_date first_name  last_name gender   hire_date  duplicated
0    10001  1953-09-02     Georgi    Facello      M  1986-06-26       False
1    10002  1964-06-02    Bezalel     Simmel      F  1985-11-21       False
2    10003  1959-12-03      Parto    Bamford      M  1986-08-28       False
3    10004  1954-05-01  Chirstian    Koblick      M  1986-12-01       False
4    10005  1955-01-21    Kyoichi   Maliniak      M  1989-09-12       False
5    10006  1953-04-20     Anneke    Preusig      F  1989-06-02       False
6    10007  1957-05-23    Tzvetan  Zielinski      F  1989-02-10       False
7    10008  1958-02-19     Saniya   Kalloufi      M  1994-09-15       False
8    10009  1952-04-19     Sumant       Peac      F  1985-02-18       False
9    10010  1963-06-01  Duangkaew   Piveteau      F  1989-08-24       False
10   10001  1953-09-02     Georgi    Facello      M  1986-06-26        True


/var/folders/pb/dhr60dvd19370mv4qdxrn34m0000gn/T/ipykernel_67521/526296586.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  emp_dup = emp.head(10); emp_dup.loc[len(emp_dup)] = emp_dup.loc[0];
/var/folders/pb/dhr60dvd19370mv4qdxrn34m0000gn/T/ipykernel_67521/25059225.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  table['duplicated'] = table.duplicated(); print(table)


emp_no  birth_date first_name  last_name gender   hire_date
0   10001  1953-09-02     Georgi    Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel     Simmel      F  1985-11-21
2   10003  1959-12-03      Parto    Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian    Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi   Maliniak      M  1989-09-12
5   10006  1953-04-20     Anneke    Preusig      F  1989-06-02
6   10007  1957-05-23    Tzvetan  Zielinski      F  1989-02-10
7   10008  1958-02-19     Saniya   Kalloufi      M  1994-09-15
8   10009  1952-04-19     Sumant       Peac      F  1985-02-18
9   10010  1963-06-01  Duangkaew   Piveteau      F  1989-08-24

# JOINS

[Back to contents](#cont) 

**<a id='INNER'>INNER</a>**

**The intersection of two tables**

In [4]:
def JOIN(table1,table2,on,how):
    return pd.merge(table1,table2,on = [on], how = how)

In [460]:
print(JOIN(emp,man,'emp_no','inner')) 
# E.g. Intersect (overlap) of employee number in employee and manager tables - checks also for null values

    emp_no  birth_date   first_name     last_name gender   hire_date dept_no  \
0   110022  1956-09-12    Margareta    Markovitch      M  1985-01-01    d001   
1   110039  1963-06-21     Vishwani      Minakawa      M  1986-04-12    d001   
2   110085  1959-10-28         Ebru         Alpin      M  1985-01-01    d002   
3   110114  1957-03-28        Isamu    Legleitner      F  1985-01-14    d002   
4   110183  1953-06-24      Shirish  Ossenbruggen      F  1985-01-01    d003   
5   110228  1958-12-02      Karsten       Sigstam      F  1985-08-04    d003   
6   110303  1956-06-08    Krassimir       Wegerle      F  1985-01-01    d004   
7   110344  1961-09-07       Rosine         Cools      F  1985-11-22    d004   
8   110386  1953-10-04         Shem        Kieras      M  1988-10-14    d004   
9   110420  1963-07-27        Oscar      Ghazalie      M  1992-02-05    d004   
10  110511  1957-07-08     DeForest      Hagimont      M  1985-01-01    d005   
11  110567  1964-04-25         Leon     

[Back to contents](#cont) 

**<a id='OUTER'>OUTER</a>**

**UNION ALL**

In [97]:
print(JOIN(emp,man,'emp_no','outer')) 
# E.g. Union of employee numers in either the employee and manager tables 

        emp_no  birth_date first_name last_name gender    hire_date dept_no  \
0        10001  1953-09-02     Georgi   Facello      M   1986-06-26     NaN   
1        10002  1964-06-02    Bezalel    Simmel      F   1985-11-21     NaN   
2        10003  1959-12-03      Parto   Bamford      M   1986-08-28     NaN   
3        10004  1954-05-01  Chirstian   Koblick      M   1986-12-01     NaN   
4        10005  1955-01-21    Kyoichi  Maliniak      M   1989-09-12     NaN   
...        ...         ...        ...       ...    ...          ...     ...   
300019  499995  1958-09-24     Dekang  Lichtner      F   1993-01-12     NaN   
300020  499996  1953-03-07       Zito      Baaz      M   1990-09-27     NaN   
300021  499997  1961-08-03    Berhard    Lenart      M   1986-04-21     NaN   
300022  499998  1956-09-05   Patricia   Breugel      M   1993-10-13     NaN   
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30;     NaN   

       from_date to_date  key  
0            NaN   

**UNION**

In [441]:
print(JOIN(emp,man,'emp_no','outer').drop_duplicates()) 

        emp_no  birth_date first_name last_name gender    hire_date dept_no  \
0        10001  1953-09-02     Georgi   Facello      M   1986-06-26     NaN   
1        10002  1964-06-02    Bezalel    Simmel      F   1985-11-21     NaN   
2        10003  1959-12-03      Parto   Bamford      M   1986-08-28     NaN   
3        10004  1954-05-01  Chirstian   Koblick      M   1986-12-01     NaN   
4        10005  1955-01-21    Kyoichi  Maliniak      M   1989-09-12     NaN   
...        ...         ...        ...       ...    ...          ...     ...   
300019  499995  1958-09-24     Dekang  Lichtner      F   1993-01-12     NaN   
300020  499996  1953-03-07       Zito      Baaz      M   1990-09-27     NaN   
300021  499997  1961-08-03    Berhard    Lenart      M   1986-04-21     NaN   
300022  499998  1956-09-05   Patricia   Breugel      M   1993-10-13     NaN   
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30;     NaN   

       from_date to_date  key  
0            NaN   

[Back to contents](#cont) 

**<a id='LR'>LEFT/RIGHT</a>**

In [ ]:
print(JOIN(emp,man,'emp_no','left')) #E.g. All records from the employee table matched with the manager table

In [55]:
print(JOIN(emp,man,'emp_no','right')) #E.g. All records from the manager table matched with the employee table

------------------------------------
table1 contains null values -  False
table2 contains null values -  False
------------------------------------
    emp_no  birth_date   first_name     last_name gender   hire_date dept_no  \
0   110022  1956-09-12    Margareta    Markovitch      M  1985-01-01    d001   
1   110039  1963-06-21     Vishwani      Minakawa      M  1986-04-12    d001   
2   110085  1959-10-28         Ebru         Alpin      M  1985-01-01    d002   
3   110114  1957-03-28        Isamu    Legleitner      F  1985-01-14    d002   
4   110183  1953-06-24      Shirish  Ossenbruggen      F  1985-01-01    d003   
5   110228  1958-12-02      Karsten       Sigstam      F  1985-08-04    d003   
6   110303  1956-06-08    Krassimir       Wegerle      F  1985-01-01    d004   
7   110344  1961-09-07       Rosine         Cools      F  1985-11-22    d004   
8   110386  1953-10-04         Shem        Kieras      M  1988-10-14    d004   
9   110420  1963-07-27        Oscar      Ghazalie   

[Back to contents](#cont) 

**<a id='MUL'>MULTIPLE</a>**

**Interactive multiple joins**  

In [57]:
def MULT_JOIN():
    t1 = str(input("table1 [e.g. man]: ")); table1 = globals()[t1]; 
    t2 = str(input("table2 [e.g. emp]: ")); table2 = globals()[t2]; 
    print("****** %s ******" %(t1))
    print(table1.head(5))
    print("****** %s ******" %(t2))     
    print(table2.head(5))    
    on = str(input("\njoin on [e.g. emp_no]: "))
    how = str(input("how [e.g. left]: "))
    tmp = JOIN(table1,table2,on,how); 
    print("\n****** %s JOIN %s ******" %(t1,t2))
    print(tmp.head(5))
    again = str(input("Join another table with these [y/n]? "))
    while again != "n": 
        l = str(input("List available tables [e.g. y/n]: "))
        if l != "n":
            %whos DataFrame
        t = str(input("table [e.g. dep]: "))
        table = globals()[t] 
        print("************ %s ************" %(t))
        print(table.head(5))
        on = str(input("join on [e.g. dept_no]: "))
        how = str(input("how [e.g. left]: "))
        tmp1 = JOIN(tmp,table,on,how)
        print("\n************ JOINED ************")
        tmp = tmp1.copy()
        print(tmp)       
        again = str(input("Join another table with these [y/n]? "))
    return tmp

E.g. you need the first and last names of all managers, their hire dates, 
the dates when they were promoted to managers and the names of their departments.

Run the following cell with replies<br>
    table1 [e.g. man]: man<br>
    table2 [e.g. emp]: emp<br>
    join on [e.g. emp_no]: emp_no<br>
    how [e.g. left]: left<br>

Join another table with these [y/n]? y<br>
    table: dep<br>
    join on: dept_no<br>
    how [e.g. left]: left<br>

Join another table with these [y/n]? y<br>
    table [e.g. dep]: titles<br>
    join on [e.g. dept_no]: emp_no<br>
    how [e.g. left]: left<br>

Join another table with these [y/n]? n

In [463]:
tmp = MULT_JOIN()
SELECT(tmp,'title',"Manager").reset_index()

table1 [e.g. man]: man
table2 [e.g. emp]: emp
****** man ******
   emp_no dept_no   from_date     to_date key
0  110022    d001  1985-01-01  1991-10-01   0
1  110039    d001  1991-10-01  9999-01-01   0
2  110085    d002  1985-01-01  1989-12-17   0
3  110114    d002  1989-12-17  9999-01-01   0
4  110183    d003  1985-01-01  1992-03-21   0
****** emp ******
   emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

join on [e.g. emp_no]: emp_no
how [e.g. left]: left

****** man JOIN emp ******
   emp_no dept_no   from_date     to_date key  birth_date first_name  \
0  110022    d001  1985-01-01  1991-10-01   0  1956-09-12  Margareta   
1  110039    d001  1991-10-01  9999-01-01   0

Join another table with these [y/n]? n


index  emp_no dept_no from_date_x    to_date_x key_x  birth_date  \
0       0  110022    d001  1985-01-01   1991-10-01     0  1956-09-12   
1       2  110039    d001  1991-10-01   9999-01-01     0  1963-06-21   
2       4  110085    d002  1985-01-01   1989-12-17     0  1959-10-28   
3       6  110114    d002  1989-12-17   9999-01-01     0  1957-03-28   
4       8  110183    d003  1985-01-01   1992-03-21     0  1953-06-24   
5      10  110228    d003  1992-03-21   9999-01-01     0  1958-12-02   
6      12  110303    d004  1985-01-01   1988-09-09     0  1956-06-08   
7      14  110344    d004  1988-09-09   1992-08-02     0  1961-09-07   
8      17  110386    d004  1992-08-02   1996-08-30     0  1953-10-04   
9      20  110420    d004  1996-08-30   9999-01-01     0  1963-07-27   
10     22  110511    d005  1985-01-01   1992-04-25     0  1957-07-08   
11     24  110567    d005  1992-04-25   9999-01-01     0  1964-04-25   
12     26  110725    d006  1985-01-01   1989-05-06     0  1961-03-14   
13     28  110765    d006  1989-05-06   1991-09-12     0  1954-05-22   
14     31  110800    d006  1991-09-12   1994-06-28     0  1963-02-07   
15     34  110854    d006  1994-06-28   9999-01-01     0  1960-08-19   
16     36  111035    d007  1985-01-01   1991-03-07     0  1962-02-24   
17     38  111133    d007  1991-03-07   9999-01-01     0  1955-03-16   
18     40  111400    d008  1985-01-01   1991-04-08     0  1959-11-09   
19     42  111534    d008  1991-04-08   9999-01-01     0  1952-06-27   
20     44  111692    d009  1985-01-01   1988-10-17     0  1954-10-05   
21     46  111784    d009  1988-10-17   1992-09-08     0  1956-06-14   
22     49  111877    d009  1992-09-08   1996-01-03     0  1962-10-18   
23     52  111939    d009  1996-01-03  9999-01-01;     0  1960-03-25   

     first_name     last_name gender   hire_date           dept_name key_y  \
0     Margareta    Markovitch      M  1985-01-01           Marketing     0   
1      Vishwani      Minakawa      M  1986-04-12           Marketing     0   
2          Ebru         Alpin      M  1985-01-01             Finance     0   
3         Isamu    Legleitner      F  1985-01-14             Finance     0   
4       Shirish  Ossenbruggen      F  1985-01-01     Human Resources     0   
5       Karsten       Sigstam      F  1985-08-04     Human Resources     0   
6     Krassimir       Wegerle      F  1985-01-01          Production     0   
7        Rosine         Cools      F  1985-11-22          Production     0   
8          Shem        Kieras      M  1988-10-14          Production     0   
9         Oscar      Ghazalie      M  1992-02-05          Production     0   
10     DeForest      Hagimont      M  1985-01-01         Development     0   
11         Leon      DasSarma      F  1986-10-21         Development     0   
12    Peternela       Onuegbe      F  1985-01-01  Quality Management     0   
13       Rutger       Hofmeyr      F  1989-01-07  Quality Management     0   
14       Sanjoy       Quadeer      F  1986-08-12  Quality Management     0   
15         Dung         Pesch      M  1989-06-09  Quality Management     0   
16  Przemyslawa     Kaelbling      M  1985-01-01               Sales     0   
17        Hauke         Zhang      M  1986-12-30               Sales     0   
18         Arie       Staelin      M  1985-01-01            Research     0   
19       Hilary        Kambil      F  1988-01-31            Research     0   
20        Tonny   Butterworth      F  1985-01-01   Customer Service;     0   
21        Marjo    Giarratana      F  1988-02-12   Customer Service;     0   
22      Xiaobin      Spinelli      F  1991-08-17   Customer Service;     0   
23      Yuchang       Weedman      M  1989-07-10   Customer Service;     0   

      title from_date_y   to_date_y  
0   Manager  1985-01-01  1991-10-01  
1   Manager  1991-10-01  9999-01-01  
2   Manager  1985-01-01  1989-12-17  
3   Manager  1989-12-17  9999-01-01  
4   Manager  1985-01-01  1992-03-21  
5   Manager  1992-03-21 

**<a id='JW'>JOIN WHERE</a>**

In [456]:
# E.g. Select all details of employees whose first name is “Margareta” and have the last name “Markovitch”.
tmp = JOIN(emp,titles,'emp_no','inner')
SELECT_WHERE(tmp,'first_name',"Margareta",'last_name',"Markovitch")

emp_no  birth_date first_name   last_name gender   hire_date  \
148135  110022  1956-09-12  Margareta  Markovitch      M  1985-01-01   
148136  110022  1956-09-12  Margareta  Markovitch      M  1985-01-01   

               title   from_date     to_date  
148135       Manager  1985-01-01  1991-10-01  
148136  Senior Staff  1991-10-01  9999-01-01

[Back to contents](#cont) 

**<a id='CJ'>CROSS JOIN</a>**

In [457]:
def CROSS_JOIN(table1,table2):
        #print(table1.head(5)); print(table2.head(5)) 
        #Create a common 'key' to cross join the two, then outer
        #key = str(input("Enter key: "))
        key = '0'
        table1['key'] = key; table2['key'] = key
        tmp = JOIN(table1,table2,'key','outer')
        del tmp['key']
        return tmp

In [ ]:
# E.g. return a list with all possible combinations between managers from the dept_manager 
# table and department number 9.
tmp = CROSS_JOIN(man,dep); print(tmp)
field = str(input("Which column [e.g. dept_no_y]? "))
value = str(input("Which value [e.g. dd009]? "))
SELECT(tmp,field,value)

     emp_no dept_no_x   from_date      to_date dept_no_y           dept_name
0    110022      d001  1985-01-01   1991-10-01      d001           Marketing
1    110022      d001  1985-01-01   1991-10-01      d002             Finance
2    110022      d001  1985-01-01   1991-10-01      d003     Human Resources
3    110022      d001  1985-01-01   1991-10-01      d004          Production
4    110022      d001  1985-01-01   1991-10-01      d005         Development
..      ...       ...         ...          ...       ...                 ...
211  111939      d009  1996-01-03  9999-01-01;      d005         Development
212  111939      d009  1996-01-03  9999-01-01;      d006  Quality Management
213  111939      d009  1996-01-03  9999-01-01;      d007               Sales
214  111939      d009  1996-01-03  9999-01-01;      d008            Research
215  111939      d009  1996-01-03  9999-01-01;      d009   Customer Service;

[216 rows x 6 columns]


[Back to contents](#cont)

**<a id='AJ'>Aggregate and JOIN</a>**

In [204]:
def AGG_JOIN(table1,table2,on,how):
    tmp = JOIN_NULL(table1,table2,on,how)
    print(tmp.head(5))
    field = str(input("Which group [e.g. gender]? "))
    value = str(input("Which quantity [e.g. salary]? "))
    action = str(input("Which function [mean, median, max or min]? "))
    out = GROUP_AGG(tmp,field,value,action); del out[on]
    return out

In [203]:
AGG_JOIN(emp,sal,'emp_no','outer') # E.g. The average, minimum or maximum salary by gender

   emp_no  birth_date first_name last_name gender   hire_date   salary  \
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26  60117.0   
1   10001  1953-09-02     Georgi   Facello      M  1986-06-26  62102.0   
2   10001  1953-09-02     Georgi   Facello      M  1986-06-26  66074.0   
3   10001  1953-09-02     Georgi   Facello      M  1986-06-26  66596.0   
4   10001  1953-09-02     Georgi   Facello      M  1986-06-26  66961.0   

    from_date     to_date  
0  1986-06-26  1987-06-26  
1  1987-06-26  1988-06-25  
2  1988-06-25  1989-06-25  
3  1989-06-25  1990-06-25  
4  1990-06-25  1991-06-25  
Which group [e.g. gender]? gender
Which quantity [e.g. salary]? salary
Which function [mean, max or min]? median


salary
gender         
F       61113.0
M       61042.0

In [428]:
# E.g. The names of all departments and calculate the average salary paid to the managers in each of them.
tmp = JOIN(man,dep,'dept_no', 'left')
tmp1 = JOIN(tmp,titles,'emp_no','left')
tmp2 = JOIN(tmp1,sal,'emp_no', 'inner');
tmp3 = tmp2[tmp2['title'] == "Manager"]
del tmp3['emp_no']
GROUP_AGG(tmp3,'dept_name','salary','mean')

salary
dept_name                       
Customer Service;   54959.672414
Development         59658.117647
Finance             70815.888889
Human Resources     58286.055556
Marketing           88371.685714
Production          56233.400000
Quality Management  67130.935484
Research            77535.181818
Sales               85738.764706

In [42]:
def DATE_RANGES(data,field1,field2):
    d1 = AGG(data,field1,'min')
    d2 = AGG(data,field1,'max')
    d3 = AGG(data,field2,'min')
    d4 = AGG(data,field2,'max')
    print(field1,'ranges over', d1,d2,' and ', field2, 'ranges over', d3,d4)

In [360]:
DATE_RANGES(tmp3,'from_date','to_date')

from_date ranges over 1985-01-01 2002-07-31  and  to_date ranges over 1986-01-01 9999-01-01


In [727]:
# E.g. Max salary after from_date = 2000-01-01
tmp4 = FILTER(tmp3,'from_date','2000-01-01','>')

In [726]:
GROUP_AGG(tmp4,'dept_name','salary','max') 

salary
dept_name                 
Customer Service;    73953
Development          74510
Finance              88443
Human Resources      79229
Marketing           108407
Production           75121
Quality Management   93193
Research            103244
Sales               101987

In [432]:
# Alternatlively
tmp = MULT_JOIN()
# table1 [e.g. man]: man
# table2 [e.g. emp]: dep
# join on [e.g. emp_no]: dept_no
# how [e.g. left]: left
# table [e.g. dep]: titles
# join on [e.g. dept_no]: emp_no
# how [e.g. left]: left
# table [e.g. dep]: sal
# join on [e.g. dept_no]: emp_no
# how [e.g. left]: inner
# Join another table with these [y/n]? n
tmp3 = tmp[tmp['title'] == "Manager"]
del tmp3['emp_no']
GROUP_AGG(tmp3,'dept_name','salary','mean')

table1 [e.g. man]: man
table2 [e.g. emp]: dep
****** man ******
   emp_no dept_no   from_date     to_date key
0  110022    d001  1985-01-01  1991-10-01   0
1  110039    d001  1991-10-01  9999-01-01   0
2  110085    d002  1985-01-01  1989-12-17   0
3  110114    d002  1989-12-17  9999-01-01   0
4  110183    d003  1985-01-01  1992-03-21   0
****** dep ******
  dept_no        dept_name key
0    d001        Marketing   0
1    d002          Finance   0
2    d003  Human Resources   0
3    d004       Production   0
4    d005      Development   0

join on [e.g. emp_no]: dept_no
how [e.g. left]: left

****** man JOIN dep ******
   emp_no dept_no   from_date     to_date key_x        dept_name key_y
0  110022    d001  1985-01-01  1991-10-01     0        Marketing     0
1  110039    d001  1991-10-01  9999-01-01     0        Marketing     0
2  110085    d002  1985-01-01  1989-12-17     0          Finance     0
3  110114    d002  1989-12-17  9999-01-01     0          Finance     0
4  110183    d003  

Join another table with these [y/n]? y
List available tables [e.g. y/n]: n
table [e.g. dep]: sal
************ sal ************
   emp_no  salary   from_date     to_date
0   10001   60117  1986-06-26  1987-06-26
1   10001   62102  1987-06-26  1988-06-25
2   10001   66074  1988-06-25  1989-06-25
3   10001   66596  1989-06-25  1990-06-25
4   10001   66961  1990-06-25  1991-06-25
join on [e.g. dept_no]: emp_no
how [e.g. left]: inner

************ JOINED ************
     emp_no dept_no from_date_x    to_date_x key_x          dept_name key_y  \
0    110022    d001  1985-01-01   1991-10-01     0          Marketing     0   
1    110022    d001  1985-01-01   1991-10-01     0          Marketing     0   
2    110022    d001  1985-01-01   1991-10-01     0          Marketing     0   
3    110022    d001  1985-01-01   1991-10-01     0          Marketing     0   
4    110022    d001  1985-01-01   1991-10-01     0          Marketing     0   
..      ...     ...         ...          ...   ...         

salary
dept_name                       
Customer Service;   54959.672414
Development         59658.117647
Finance             70815.888889
Human Resources     58286.055556
Marketing           88371.685714
Production          56233.400000
Quality Management  67130.935484
Research            77535.181818
Sales               85738.764706

In [737]:
# E.g. what is the gender of the managers?
tmp = JOIN(man,emp,'emp_no','left')
GROUP_COUNT(tmp,'gender')

gender  Count
0      F     13
1      M     11

[Back to contents](#cont)

**<a id='CAS'>CASE</a>**

**Example 1: Add a new field (named Position) to the employee table stating whether a manager or not**

In [548]:
def CASE_1(table1,table2,field,new,default,case):
    table = table1.copy()
    table[new] = default
    table.loc[table[field].isin(table2[field]), new] = case
    return table

In [550]:
CASE_1(emp,man,'emp_no','Position','Not manager','Manager')

emp_no  birth_date first_name last_name gender    hire_date  \
0        10001  1953-09-02     Georgi   Facello      M   1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F   1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M   1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M   1986-12-01   
4        10005  1955-01-21    Kyoichi  Maliniak      M   1989-09-12   
...        ...         ...        ...       ...    ...          ...   
300019  499995  1958-09-24     Dekang  Lichtner      F   1993-01-12   
300020  499996  1953-03-07       Zito      Baaz      M   1990-09-27   
300021  499997  1961-08-03    Berhard    Lenart      M   1986-04-21   
300022  499998  1956-09-05   Patricia   Breugel      M   1993-10-13   
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30;   

           Position  
0       Not manager  
1       Not manager  
2       Not manager  
3       Not manager  
4       Not manager  
...             ...  
300019  Not manager  
300020  Not manager  
300021  Not manager  
300022  Not manager  
300023  Not manager  

[300024 rows x 7 columns]

[Back to contents](#cont)

**Example 2: List managers' total salary increase**

In [1051]:
tmp = JOIN(man,emp,'emp_no','left')
tmp1 = JOIN(tmp,sal,'emp_no','left')
IDs = tmp1['emp_no'].unique()
columns = ['emp_no','dept_no','first_name','last_name','diff']

dfs = []
for i in range (0,len(IDs)):
    tmp = tmp1[tmp1['emp_no'] == IDs[i]]
    tmp.sort_values(by=['salary'], ascending=True);
    tmp['diff'] = tmp['salary'].iloc[-1] - tmp['salary'].iloc[0]
    tmp = tmp[columns].reset_index(); del tmp['index']
    try:
        tmp2=tmp.iloc[[0]]; # DOUBLE BRACKETS FORCE SINGLE LINE
        dfs.append(tmp2)
    except IndexError: 
        pass
out = pd.concat(dfs)
out

# if a certain manager has a raise higher than $30 000, we will return, 
# "Salary was raised by more than $30 000". If, instead, it’s between $20 000 and $30 000, 
# we would like to see, "Salary was raised by more than $20 000" but less than $30 000 displayed. 
# Finally, if none of the conditions were met the query would return, "Salary was raised by less than $20 000".

d1 = 30000; d2 = 20000
print('\n===================== Splitting at %1.0fk and %1.0fk =====================' %(d2/1000,d1/1000))
out.loc[out['diff'] > d1, 'Salary increase'] = 'More than %1.0fk' %(d1/1000)
out.loc[out['diff'] < d2, 'Salary increase'] = 'Less than %1.0fk' %(d2/1000)
out.loc[(out['diff'] >= d2) & (out['diff'] <= d1), 'Salary increase'] = 'By between %1.0fk and %1.0fk' %(d2/1000,d1/1000)
del out['diff'] 
out



===================== Splitting at 20k and 30k =====================


emp_no dept_no   first_name     last_name         Salary increase
0  110022    d001    Margareta    Markovitch           More than 30k
0  110039    d001     Vishwani      Minakawa           More than 30k
0  110085    d002         Ebru         Alpin  By between 20k and 30k
0  110114    d002        Isamu    Legleitner           More than 30k
0  110183    d003      Shirish  Ossenbruggen           More than 30k
0  110228    d003      Karsten       Sigstam  By between 20k and 30k
0  110303    d004    Krassimir       Wegerle           More than 30k
0  110344    d004       Rosine         Cools  By between 20k and 30k
0  110386    d004         Shem        Kieras  By between 20k and 30k
0  110420    d004        Oscar      Ghazalie           Less than 20k
0  110511    d005     DeForest      Hagimont  By between 20k and 30k
0  110567    d005         Leon      DasSarma           More than 30k
0  110725    d006    Peternela       Onuegbe           More than 30k
0  110765    d006       Rutger       Hofmeyr           More than 30k
0  110800    d006       Sanjoy       Quadeer  By between 20k and 30k
0  110854    d006         Dung         Pesch  By between 20k and 30k
0  111035    d007  Przemyslawa     Kaelbling  By between 20k and 30k
0  111133    d007        Hauke         Zhang           More than 30k
0  111400    d008         Arie       Staelin           More than 30k
0  111534    d008       Hilary        Kambil           More than 30k
0  111692    d009        Tonny   Butterworth           More than 30k
0  111784    d009        Marjo    Giarratana           Less than 20k
0  111877    d009      Xiaobin      Spinelli           Less than 20k
0  111939    d009      Yuchang       Weedman           Less than 20k

[Back to contents](#cont)

**Example 3: Obtain a table containing the employee number, first name, and last name of all employees with a number higher than 109990. Create a column indicating whether this employee is also a manager, according to the dept_manager table, or a regular employee.**


In [713]:
tmp = CASE_1(emp,man,'emp_no','Position','Not manager','Manager')
FILTER(tmp,'emp_no',109990,'>').reset_index()

index  emp_no  birth_date   first_name last_name gender    hire_date  \
0        99990  109991  1954-10-30        Jinxi   Reistad      M   1987-01-13   
1        99991  109992  1955-03-21       Cheong   Heering      F   1993-03-25   
2        99992  109993  1953-10-19        Danel  Furudate      M   1986-11-04   
3        99993  109994  1956-10-24    Guangming  Takkinen      M   1986-05-24   
4        99994  109995  1960-02-11  Constantijn     Anido      F   1986-06-11   
...        ...     ...         ...          ...       ...    ...          ...   
200029  300019  499995  1958-09-24       Dekang  Lichtner      F   1993-01-12   
200030  300020  499996  1953-03-07         Zito      Baaz      M   1990-09-27   
200031  300021  499997  1961-08-03      Berhard    Lenart      M   1986-04-21   
200032  300022  499998  1956-09-05     Patricia   Breugel      M   1993-10-13   
200033  300023  499999  1958-05-01       Sachin   Tsukuda      M  1997-11-30;   

           Position  
0       Not manager  
1       Not manager  
2       Not manager  
3       Not manager  
4       Not manager  
...             ...  
200029  Not manager  
200030  Not manager  
200031  Not manager  
200032  Not manager  
200033  Not manager  

[200034 rows x 8 columns]

[Back to contents](#cont)

**Example 4: Extract the employee number, first name, and last name of the first 100 employees. Add column stating if the  employee is still working in the company.**

In [734]:
tmp = emp[['emp_no','first_name','last_name']]
tmp1= JOIN(tmp,dept_emp,'emp_no','left')
tmp2 = tmp1.head(100)
Dict = {"9999-01-01": 'Not an employee anymore'}
tmp2['current_employee']= tmp2['to_date'].map(Dict)
tmp2['current_employee'] = tmp2['current_employee'].replace({np.nan: 'Is still employed'})
tmp2

emp_no first_name     last_name dept_no   from_date     to_date  \
0    10001     Georgi       Facello    d005  1986-06-26  9999-01-01   
1    10002    Bezalel        Simmel    d007  1996-08-03  9999-01-01   
2    10003      Parto       Bamford    d004  1995-12-03  9999-01-01   
3    10004  Chirstian       Koblick    d004  1986-12-01  9999-01-01   
4    10005    Kyoichi      Maliniak    d003  1989-09-12  9999-01-01   
..     ...        ...           ...     ...         ...         ...   
95   10088   Jungsoon      Syrzycki    d007  1988-09-02  1992-03-21   
96   10088   Jungsoon      Syrzycki    d009  1992-03-21  9999-01-01   
97   10089  Sudharsan  Flasterstein    d007  1989-01-10  9999-01-01   
98   10090     Kendra       Hofting    d005  1986-03-14  1999-05-07   
99   10091    Amabile       Gomatam    d005  1992-11-18  9999-01-01   

           current_employee  
0   Not an employee anymore  
1   Not an employee anymore  
2   Not an employee anymore  
3   Not an employee anymore  
4   Not an employee anymore  
..                      ...  
95        Is still employed  
96  Not an employee anymore  
97  Not an employee anymore  
98        Is still employed  
99  Not an employee anymore  

[100 rows x 7 columns]

[Back to contents](#cont)

# <a id='WF'>Window Functions</a>

| | | |
| :---                            | :---                   | :--- |
|**[ROW NUMBER](#RN)** | **[PARTITION BY](#PB)** | **[RANK](#RANK)** |
| **[LAG and LEAD](#LaL)**|**[Aggregate Functions](#WAF)** |**[Common Table Expressions](#CTE)** |

**<a id='RN'>ROW_NUMBER</a>**

**Example 1: Write a query that upon execution, assigns a row number to all managers we have information for in the "employees" database (regardless of their department).**

**Let the numbering disregard the department the managers have worked in. Also, let it start from the value of 1. Assign that value to the manager with the lowest employee number.**

In [741]:
tmp = JOIN(man,emp,'emp_no','left').reset_index();
tmp = tmp[['emp_no','dept_no','first_name','last_name']]
tmp["row_number"] = tmp.index+1
tmp

emp_no dept_no   first_name     last_name  row_number
0   110022    d001    Margareta    Markovitch           1
1   110039    d001     Vishwani      Minakawa           2
2   110085    d002         Ebru         Alpin           3
3   110114    d002        Isamu    Legleitner           4
4   110183    d003      Shirish  Ossenbruggen           5
5   110228    d003      Karsten       Sigstam           6
6   110303    d004    Krassimir       Wegerle           7
7   110344    d004       Rosine         Cools           8
8   110386    d004         Shem        Kieras           9
9   110420    d004        Oscar      Ghazalie          10
10  110511    d005     DeForest      Hagimont          11
11  110567    d005         Leon      DasSarma          12
12  110725    d006    Peternela       Onuegbe          13
13  110765    d006       Rutger       Hofmeyr          14
14  110800    d006       Sanjoy       Quadeer          15
15  110854    d006         Dung         Pesch          16
16  111035    d007  Przemyslawa     Kaelbling          17
17  111133    d007        Hauke         Zhang          18
18  111400    d008         Arie       Staelin          19
19  111534    d008       Hilary        Kambil          20
20  111692    d009        Tonny   Butterworth          21
21  111784    d009        Marjo    Giarratana          22
22  111877    d009      Xiaobin      Spinelli          23
23  111939    d009      Yuchang       Weedman          24

[Back to contents](#cont)

**Example 2: Get the employee number of each manager**

In [106]:
tmp = JOIN(man,emp,'emp_no','left')
tmp1= tmp.groupby("emp_no").rank(method="first", ascending=True)
tmp["row_number"] = tmp1['dept_no'].astype(int) # DOESN'T MATTER WHICH FIELD
tmp1 = tmp[tmp["row_number"]==1].reset_index()
tmp1

index  emp_no dept_no   from_date      to_date  birth_date   first_name  \
0       0  110022    d001  1985-01-01   1991-10-01  1956-09-12    Margareta   
1       1  110039    d001  1991-10-01   9999-01-01  1963-06-21     Vishwani   
2       2  110085    d002  1985-01-01   1989-12-17  1959-10-28         Ebru   
3       3  110114    d002  1989-12-17   9999-01-01  1957-03-28        Isamu   
4       4  110183    d003  1985-01-01   1992-03-21  1953-06-24      Shirish   
5       5  110228    d003  1992-03-21   9999-01-01  1958-12-02      Karsten   
6       6  110303    d004  1985-01-01   1988-09-09  1956-06-08    Krassimir   
7       7  110344    d004  1988-09-09   1992-08-02  1961-09-07       Rosine   
8       8  110386    d004  1992-08-02   1996-08-30  1953-10-04         Shem   
9       9  110420    d004  1996-08-30   9999-01-01  1963-07-27        Oscar   
10     10  110511    d005  1985-01-01   1992-04-25  1957-07-08     DeForest   
11     11  110567    d005  1992-04-25   9999-01-01  1964-04-25         Leon   
12     12  110725    d006  1985-01-01   1989-05-06  1961-03-14    Peternela   
13     13  110765    d006  1989-05-06   1991-09-12  1954-05-22       Rutger   
14     14  110800    d006  1991-09-12   1994-06-28  1963-02-07       Sanjoy   
15     15  110854    d006  1994-06-28   9999-01-01  1960-08-19         Dung   
16     16  111035    d007  1985-01-01   1991-03-07  1962-02-24  Przemyslawa   
17     17  111133    d007  1991-03-07   9999-01-01  1955-03-16        Hauke   
18     18  111400    d008  1985-01-01   1991-04-08  1959-11-09         Arie   
19     19  111534    d008  1991-04-08   9999-01-01  1952-06-27       Hilary   
20     20  111692    d009  1985-01-01   1988-10-17  1954-10-05        Tonny   
21     21  111784    d009  1988-10-17   1992-09-08  1956-06-14        Marjo   
22     22  111877    d009  1992-09-08   1996-01-03  1962-10-18      Xiaobin   
23     23  111939    d009  1996-01-03  9999-01-01;  1960-03-25      Yuchang   

       last_name gender   hire_date  row_number  
0     Markovitch      M  1985-01-01           1  
1       Minakawa      M  1986-04-12           1  
2          Alpin      M  1985-01-01           1  
3     Legleitner      F  1985-01-14           1  
4   Ossenbruggen      F  1985-01-01           1  
5        Sigstam      F  1985-08-04           1  
6        Wegerle      F  1985-01-01           1  
7          Cools      F  1985-11-22           1  
8         Kieras      M  1988-10-14           1  
9       Ghazalie      M  1992-02-05           1  
10      Hagimont      M  1985-01-01           1  
11      DasSarma      F  1986-10-21           1  
12       Onuegbe      F  1985-01-01           1  
13       Hofmeyr      F  1989-01-07           1  
14       Quadeer      F  1986-08-12           1  
15         Pesch      M  1989-06-09           1  
16     Kaelbling      M  1985-01-01           1  
17         Zhang      M  1986-12-30           1  
18       Staelin      M  1985-01-01           1  
19        Kambil      F  1988-01-31           1  
20   Butterworth      F  1985-01-01           1  
21    Giarratana      F  1988-02-12           1  
22      Spinelli      F  1991-08-17           1  
23       Weedman      M  1989-07-10           1

[Back to contents](#cont)

**Example 3: Compare the salaries between departments**

In [92]:
#DATE_RANGES(sal,'from_date','to_date')
sal_fixed = sal[sal['to_date'] < '9999-01-01'] # These employees have left
#DATE_RANGES(sal_fixed,'from_date','to_date') # Latest is now 2002-08-01 # CHOOSE DATE RANGES 
tmp = JOIN(emp,sal_fixed,'emp_no','left')
tmp1 = JOIN(tmp,dept_emp,'emp_no','left')
tmp = tmp1.dropna(subset = ['salary'])
tmp1 = JOIN(tmp,dep,'dept_no','left')
tmp1 = tmp1[['emp_no','first_name','last_name','salary','dept_no','dept_name','from_date_x','to_date_x']]

tmp = GROUP_AGG(tmp1,'dept_no','salary','min')
tmp = tmp.rename(columns={'salary': 'Min salary'});# print(tmp)

tmp2 = GROUP_AGG(tmp1,'dept_no','salary','mean'); 
tmp2 = tmp2.rename(columns={'salary': 'Ave salary'});
tmp2['Ave salary'] = round(tmp2['Ave salary'],0)
tmp['Ave salary'] = tmp2['Ave salary'] 

tmp3 = GROUP_AGG(tmp1,'dept_no','salary','max');
tmp3 = tmp3.rename(columns={'salary':'Max salary'})
out = JOIN(tmp,tmp3,'dept_no', 'left')
del out['emp_no_x']; del out['emp_no_y']
out2 = JOIN(out,dep,'dept_no', 'left')
print("\n=============== Over all dates salaries in $000 are ===============\n") 
print(out2)

###### VERSION WITH HISTOGRAMS - UNCOMMENT AT BOTTOM TO RUN ###
def plot_histos(bins,date1,date2):
    import matplotlib.pyplot as plt
    
    tmp = SELECT_BETWEEN(tmp1,'to_date_x',date1,date2,'y'); # FILTER TIME RANGE AS WHOLE SET TAKES A WHILE
    depts = np.sort(tmp['dept_no'].unique())
    
    print("\n=============== Over %s to %s salaries in $000 are ===============\n" %(date1,date2))
    for i in range (0,len(depts)): 
        data = tmp[tmp['dept_no'] == depts[i]]
        dept_name = data['dept_name'].iloc[0]
        print("For %s [%s], %d staff -  salary: mean = %1.3f median = %1.3f max = %1.3f min = %1.3f" 
              %(depts[i],dept_name, len(data), 1e-3*data['salary'].mean(), 1e-3*data['salary'].median(),
                1e-3*max(data['salary']),1e-3*min(data['salary'])))
    
    font = 14
    colors = ['r','orange','gold','g','b','indigo','blueviolet','dimgrey','slategrey']    

    for i in range (0,len(depts)): 
        plt.rcParams.update({'font.size': font})
        plt.figure(figsize=(5,4))
        ax = plt.gca()
        ax.tick_params(direction='in', length=6, width=1.5, which='major')
        ax.tick_params(direction='in', length=3, width=1.5, which='minor')
        plt.setp(ax.spines.values(),linewidth=2)
        
        data = tmp[tmp['dept_no'] == depts[i]]
        dept_name = data['dept_name'].iloc[0]
        plt.ylabel('No. in %s' %(dept_name), size=font)
        plt.xlabel("Salaries in Dept %s [$000]" %(depts[i]), size=font)
    
        x1 = 1e-3*min(data['salary']); x2 = 1e-3*max(data['salary'])
        ax.hist(1e-3*data['salary'], bins = bins, color="w", edgecolor=colors[i], linewidth=3);
        y1,y2 = ax.get_ylim(); y2 = 1.1*y2 
        ax.set_ylim(x1,x2); ax.set_ylim(y1,y2)
        xpos = x2-(x2-x1)/4; ypos = y2-(y2-y1)/12; yskip = (y2-y1)/12;
        mean = 1e-3*np.mean(data['salary']); std = 1e-3*np.std(data['salary'])
        plt.text(xpos,ypos,"\u03BC = %1.2fk" %(mean), fontsize = 0.8*font, c = 'k') 
        plt.text(xpos,ypos-yskip,"\u03C3  = %1.2fk" %(std), fontsize = 0.8*font, c = 'k') 
        plt.tight_layout()
        plt.show()
#plot_histos(10,'2002-07-01','2002-08-01') # Uncomment for histograms of each department


=============== Over all dates salaries in $000 are ===============

  dept_no  Min salary  Ave salary  Max salary           dept_name
0    d001     39127.0     71160.0    142753.0           Marketing
1    d002     38812.0     69418.0    133230.0             Finance
2    d003     38735.0     54600.0    120933.0     Human Resources
3    d004     38836.0     58781.0    128773.0          Production
4    d005     38850.0     58748.0    140142.0         Development
5    d006     38928.0     56539.0    122376.0  Quality Management
6    d007     39427.0     80013.0    157821.0               Sales
7    d008     38851.0     59104.0    122121.0            Research
8    d009     38836.0     57999.0    144088.0   Customer Service;


See https://github.com/steviecurran/two-sample to apply AB testing between these

[Back to contents](#cont)

**<a id='PB'>PARTITION BY</a>**

**Example 1: Retrieve the employee number and the lowest contract salary value for all managers.**

In [113]:
tmp = JOIN(man,sal,'emp_no','left')
tmp1 = JOIN(tmp,emp,'emp_no','left'); #print(tmp1)
tmp1 = tmp1[['emp_no','first_name','last_name','dept_no','salary']]

def ID_split(data,ID,field,rank,label):
    dfs = []
    IDs = tmp1['emp_no'].unique()
    for i in range (0,len(IDs)):
        tmp= data[data[ID] == IDs[i]]
        tmp = tmp.sort_values(by=[field], ascending=False).reset_index();
        tmp[label] = tmp[field].iloc[rank-1] #min(tmp[field]); 
        try:
            tmp2=tmp.iloc[[0]]; 
            dfs.append(tmp2)
        except IndexError: 
            pass
        #del tmp2[field]
        del tmp['index']
    out = pd.concat(dfs)
    return out
ID_split(tmp1,'emp_no','salary',0,'min_sal')

index  emp_no   first_name     last_name dept_no  salary  min_sal
0     17  110022    Margareta    Markovitch    d001  108407    71166
0     34  110039     Vishwani      Minakawa    d001  106491    69941
0     52  110085         Ebru         Alpin    d002   88443    60026
0     70  110114        Isamu    Legleitner    d002   83457    52070
0     88  110183      Shirish  Ossenbruggen    d003   79229    48291
0    106  110228      Karsten       Sigstam    d003   65400    40000
0    124  110303    Krassimir       Wegerle    d004   72583    42093
0    141  110344       Rosine         Cools    d004   75121    50365
0    155  110386         Shem        Kieras    d004   66995    40000
0    166  110420        Oscar      Ghazalie    d004   56654    40000
0    180  110511     DeForest      Hagimont    d005   74056    48626
0    200  110567         Leon      DasSarma    d005   74510    40000
0    218  110725    Peternela       Onuegbe    d006   93193    61357
0    232  110765       Rutger       Hofmeyr    d006   72432    40055
0    248  110800       Sanjoy       Quadeer    d006   77626    50457
0    262  110854         Dung         Pesch    d006   72876    45169
0    280  111035  Przemyslawa     Kaelbling    d007   95873    71612
0    296  111133        Hauke         Zhang    d007  101987    70787
0    314  111400         Arie       Staelin    d008  103244    72446
0    329  111534       Hilary        Kambil    d008   79393    48077
0    347  111692        Tonny   Butterworth    d009   73953    40000
0    362  111784        Marjo    Giarratana    d009   58189    40000
0    373  111877      Xiaobin      Spinelli    d009   71406    53846
0    386  111939      Yuchang       Weedman    d009   58968    40000

**Example 2: Retrieve the employee number and the highest contract salary value for all managers.**

In [109]:
ID_split(tmp1,'emp_no','salary',1,'max_sal')

index  emp_no   first_name     last_name dept_no  max_sal
0     17  110022    Margareta    Markovitch    d001   108407
0     34  110039     Vishwani      Minakawa    d001   106491
0     52  110085         Ebru         Alpin    d002    88443
0     70  110114        Isamu    Legleitner    d002    83457
0     88  110183      Shirish  Ossenbruggen    d003    79229
0    106  110228      Karsten       Sigstam    d003    65400
0    124  110303    Krassimir       Wegerle    d004    72583
0    141  110344       Rosine         Cools    d004    75121
0    155  110386         Shem        Kieras    d004    66995
0    166  110420        Oscar      Ghazalie    d004    56654
0    180  110511     DeForest      Hagimont    d005    74056
0    200  110567         Leon      DasSarma    d005    74510
0    218  110725    Peternela       Onuegbe    d006    93193
0    232  110765       Rutger       Hofmeyr    d006    72432
0    248  110800       Sanjoy       Quadeer    d006    77626
0    262  110854         Dung         Pesch    d006    72876
0    280  111035  Przemyslawa     Kaelbling    d007    95873
0    296  111133        Hauke         Zhang    d007   101987
0    314  111400         Arie       Staelin    d008   103244
0    329  111534       Hilary        Kambil    d008    79393
0    347  111692        Tonny   Butterworth    d009    73953
0    362  111784        Marjo    Giarratana    d009    58189
0    373  111877      Xiaobin      Spinelli    d009    71406
0    386  111939      Yuchang       Weedman    d009    58968

**Example 3: Retrieve the employee number and the third-highest contract salary value
for all managers.**

In [110]:
ID_split(tmp1,'emp_no','salary',3,'3rd highest')

index  emp_no   first_name     last_name dept_no  3rd highest
0     17  110022    Margareta    Markovitch    d001       100592
0     34  110039     Vishwani      Minakawa    d001       101901
0     52  110085         Ebru         Alpin    d002        85394
0     70  110114        Isamu    Legleitner    d002        80423
0     88  110183      Shirish  Ossenbruggen    d003        74584
0    106  110228      Karsten       Sigstam    d003        62276
0    124  110303    Krassimir       Wegerle    d004        68483
0    141  110344       Rosine         Cools    d004        73754
0    155  110386         Shem        Kieras    d004        61207
0    166  110420        Oscar      Ghazalie    d004        53773
0    180  110511     DeForest      Hagimont    d005        73350
0    200  110567         Leon      DasSarma    d005        66307
0    218  110725    Peternela       Onuegbe    d006        91012
0    232  110765       Rutger       Hofmeyr    d006        65921
0    248  110800       Sanjoy       Quadeer    d006        77088
0    262  110854         Dung         Pesch    d006        69035
0    280  111035  Przemyslawa     Kaelbling    d007        93765
0    296  111133        Hauke         Zhang    d007        99072
0    314  111400         Arie       Staelin    d008        99971
0    329  111534       Hilary        Kambil    d008        78032
0    347  111692        Tonny   Butterworth    d009        72421
0    362  111784        Marjo    Giarratana    d009        54187
0    373  111877      Xiaobin      Spinelli    d009        67470
0    386  111939      Yuchang       Weedman    d009        58698

[Back to contents](#cont)

**<a id='RANK'>RANK</a>**

**Example 1: Order and number all contract salary values of employee 10002 from
highest to lowest. Store the row numbers in a third column.**

In [1101]:
tmp = JOIN(emp,sal,'emp_no','left'); #print(tmp)
tmp1 = SELECT(tmp,'emp_no',10002); #print(tmp1.columns)
tmp1 = tmp1[['emp_no', 'salary']]
tmp1 = tmp1.sort_values(by=['salary'], ascending=False).reset_index();#print(tmp1)
del tmp1['index']
tmp1['order no'] = tmp1.index +1
tmp1

emp_no   salary  order no
0   10002  72527.0         1
1   10002  71963.0         2
2   10002  69366.0         3
3   10002  67534.0         4
4   10002  65909.0         5
5   10002  65828.0         6

**Example 2: Rank the contract salary values
from highest to lowest for employees 10001, 10002, 10003, 10004, 10005, and 10006. Every row in the desired output should contain an employee number obtained from the employees table and a salary value obtained from the salaries table. Additionally, include the salary ranking values
between the two columns**

In [11]:
#pd.set_option('display.max_rows', None)
tmp = JOIN(emp,sal,'emp_no','left')
IDs = []
for i in range (10001,10007):
    IDs.append(i)
#print(IDs); print(tmp); print(tmp.columns)
tmp1 = tmp[['emp_no','salary']]
def rank_by(data,ID,field,asc):
    dfs = []
    for i in range (0,len(IDs)):
        tmp= data[data[ID] == IDs[i]]
        tmp = tmp.sort_values(by=[field], ascending=asc).reset_index()
        tmp['rank'] = tmp.index +1
        dfs.append(tmp)
    out = pd.concat(dfs)
    del out['index']
    return out
rank_by(tmp1,'emp_no','salary',False)

emp_no   salary  rank
0    10001  88958.0     1
1    10001  85112.0     2
2    10001  85097.0     3
3    10001  84917.0     4
4    10001  81097.0     5
..     ...      ...   ...
7    10006  47518.0     8
8    10006  45844.0     9
9    10006  42629.0    10
10   10006  42085.0    11
11   10006  40000.0    12

[71 rows x 3 columns]

[Back to contents](#cont)

**Example 3: Rank the contract salary
values from highest to lowest for employees 10001, 10002, and 10003.**

**Every row in the desired output should contain the relevant employee number
and the hire date from the employees table, as well as the relevant salary value and the
start date from the salaries table. Additionally, include the salaryranking values.**

**Retrieve only data for contracts that have started prior to 2000. Sort your data by the
emplyee number in ascending order.**

In [114]:
tmp = JOIN(emp,sal,'emp_no','left')
tmp1 = FILTER(tmp,'from_date','2000-01-01','<') 
IDs = [10001,10002,10003]
#print(tmp.columns)
tmp = tmp1[['emp_no','salary', 'hire_date','from_date']]
rank_by(tmp,'emp_no','salary',False)

emp_no   salary   hire_date   from_date  rank
0    10001  84917.0  1986-06-26  1999-06-23     1
1    10001  81097.0  1986-06-26  1998-06-23     2
2    10001  81025.0  1986-06-26  1997-06-23     3
3    10001  80013.0  1986-06-26  1996-06-23     4
4    10001  76884.0  1986-06-26  1995-06-24     5
5    10001  75994.0  1986-06-26  1994-06-24     6
6    10001  75286.0  1986-06-26  1993-06-24     7
7    10001  74333.0  1986-06-26  1992-06-24     8
8    10001  71046.0  1986-06-26  1991-06-25     9
9    10001  66961.0  1986-06-26  1990-06-25    10
10   10001  66596.0  1986-06-26  1989-06-25    11
11   10001  66074.0  1986-06-26  1988-06-25    12
12   10001  62102.0  1986-06-26  1987-06-26    13
13   10001  60117.0  1986-06-26  1986-06-26    14
0    10002  69366.0  1985-11-21  1999-08-03     1
1    10002  67534.0  1985-11-21  1998-08-03     2
2    10002  65909.0  1985-11-21  1997-08-03     3
3    10002  65828.0  1985-11-21  1996-08-03     4
0    10003  43636.0  1986-08-28  1998-12-02     1
1    10003  43616.0  1986-08-28  1996-12-02     2
2    10003  43478.0  1986-08-28  1999-12-02     3
3    10003  43466.0  1986-08-28  1997-12-02     4
4    10003  40006.0  1986-08-28  1995-12-03     5

[Back to contents](#cont)

**<a id='LaL'>LAG and LEAD</a>**

**Example 1: For employees with employee numbers between 10003 and 10008, inclusive, and their salary contracts with values less than $70 000, retrieve the following data from the salaries table:<br>
• employee number<br>
• salary<br>
• previous salary<br>
• next salary<br>
• the difference between the current salary of a certain employee and their previous salary<br>
• the difference between the next salary of a certain employee and their current salary<br>**

In [46]:
tmp = JOIN(emp,sal,'emp_no','left')
tmp1 = FILTER(tmp,'salary',70000,'<'); #print(tmp1)
IDs = []
for i in range (10003,10009):
    IDs.append(i)
tmp = rank_by(tmp1,'emp_no','salary',True)
tmp = tmp[['emp_no','salary']]
tmp['previous'] = tmp.salary.shift(1)
tmp['next'] = tmp.salary.shift(-1)
tmp['$\Delta$ previous'] = tmp['salary'] - tmp['previous']
tmp['$\Delta$ next'] = tmp['next'] - tmp['salary']
tmp

emp_no   salary  previous     next  $\Delta$ previous  $\Delta$ next
0    10003  40006.0       NaN  43311.0                NaN         3305.0
1    10003  43311.0   40006.0  43466.0             3305.0          155.0
2    10003  43466.0   43311.0  43478.0              155.0           12.0
3    10003  43478.0   43466.0  43616.0               12.0          138.0
4    10003  43616.0   43478.0  43636.0              138.0           20.0
5    10003  43636.0   43616.0  43699.0               20.0           63.0
6    10003  43699.0   43636.0  40054.0               63.0        -3645.0
0    10004  40054.0   43699.0  42283.0            -3645.0         2229.0
1    10004  42283.0   40054.0  42542.0             2229.0          259.0
2    10004  42542.0   42283.0  46065.0              259.0         3523.0
3    10004  46065.0   42542.0  48271.0             3523.0         2206.0
4    10004  48271.0   46065.0  50594.0             2206.0         2323.0
5    10004  50594.0   48271.0  52119.0             2323.0         1525.0
6    10004  52119.0   50594.0  54693.0             1525.0         2574.0
7    10004  54693.0   52119.0  58326.0             2574.0         3633.0
8    10004  58326.0   54693.0  60770.0             3633.0         2444.0
9    10004  60770.0   58326.0  62566.0             2444.0         1796.0
10   10004  62566.0   60770.0  64340.0             1796.0         1774.0
11   10004  64340.0   62566.0  67096.0             1774.0         2756.0
12   10004  67096.0   64340.0  69722.0             2756.0         2626.0
13   10004  69722.0   67096.0  40000.0             2626.0       -29722.0
0    10006  40000.0   69722.0  42085.0           -29722.0         2085.0
1    10006  42085.0   40000.0  42629.0             2085.0          544.0
2    10006  42629.0   42085.0  45844.0              544.0         3215.0
3    10006  45844.0   42629.0  47518.0             3215.0         1674.0
4    10006  47518.0   45844.0  47917.0             1674.0          399.0
5    10006  47917.0   47518.0  52255.0              399.0         4338.0
6    10006  52255.0   47917.0  53747.0             4338.0         1492.0
7    10006  53747.0   52255.0  56032.0             1492.0         2285.0
8    10006  56032.0   53747.0  58299.0             2285.0         2267.0
9    10006  58299.0   56032.0  59755.0             2267.0         1456.0
10   10006  59755.0   58299.0  60098.0             1456.0          343.0
11   10006  60098.0   59755.0  56724.0              343.0        -3374.0
0    10007  56724.0   60098.0  60740.0            -3374.0         4016.0
1    10007  60740.0   56724.0  62745.0             4016.0         2005.0
2    10007  62745.0   60740.0  63208.0             2005.0          463.0
3    10007  63208.0   62745.0  63475.0              463.0          267.0
4    10007  63475.0   63208.0  64563.0              267.0         1088.0
5    10007  64563.0   63475.0  68833.0             1088.0         4270.0
6    10007  68833.0   64563.0  46671.0             4270.0       -22162.0
0    10008  46671.0   68833.0  48584.0           -22162.0         1913.0
1    10008  48584.0   46671.0  52668.0             1913.0         4084.0
2    10008  52668.0   48584.0      NaN             4084.0            NaN

[Back to contents](#cont)

**Example 2: Retrieve the following data from the salaries table:<br>
• employee number<br>
• salary<br>
• the salary value of three contracts prior to the given employee contract salary value, if applicable<br>
• the salary value of three contracts after the given employee contract salary value, if applicable<br>
Retrieve only the first one hundred rows of data.**


In [56]:
tmp = JOIN(emp,sal,'emp_no','left')
tmp1 = tmp.head(100)
IDs = tmp1['emp_no'].unique() # NEED FOR rank_by FUNCTION
tmp = rank_by(tmp1,'emp_no','salary',True)
tmp = tmp[['emp_no','salary']]
no = 3
for i in range(1,no):
    tmp['%d_previous' %(no)] = tmp.salary.shift(no)
    tmp['%d_next' %(no)] = tmp.salary.shift(-no)
tmp

emp_no   salary  3_previous   3_next
0    10001  60117.0         NaN  66596.0
1    10001  62102.0         NaN  66961.0
2    10001  66074.0         NaN  71046.0
3    10001  66596.0     60117.0  74333.0
4    10001  66961.0     62102.0  75286.0
..     ...      ...         ...      ...
7    10009  74612.0     69042.0  80944.0
8    10009  76518.0     70889.0  82507.0
9    10009  78335.0     71434.0      NaN
10   10009  80944.0     74612.0      NaN
11   10009  82507.0     76518.0      NaN

[100 rows x 4 columns]

[Back to contents](#cont)

**<a id='WAF'>Aggregate Functions</a>**

**Example 1: Retrieve the employee number, salary value, start date  and end date of the latest contract of all employees according to the data stored in the salaries table.**

In [8]:
#ID_split(sal,'emp_no','from_date',1,'latest'); # TAKES WAY TO LONG, SO USE
latest_sals = GROUP_AGG(sal,'emp_no','from_date','max')
#latest_sals = latest_sals.rename(columns={'salary':'Latest salary'})
latest_sals

salary
emp_no        
10001    88958
10002    72527
10003    43699
10004    74057
10005    94692
...        ...
201767   47097
201768   40000
201769   62287
201770   85571
201771   66792

[101796 rows x 1 columns]

**Example 2: Join the salaries and dept_emp tables ON employee number, as well as the dept_emp and departments tables ON department number to select the department number from the dept_emp table, the department name from the departments table, and the salary from the salaries table. Use this subset in an outer query to retrieve:<br>
• the distinct department numbers<br>
• the relevant department name<br>
• the smallest salary value recorded for the given department<br>
• the highest salary value recorded for the given department<br>
• the average salary value recorded for the given department, rounded to the nearest dollar<br>
Sort the final output by department number in ascending order.**

In [14]:
tmp = JOIN(dept_emp,dep,'dept_no','left')
#tmp1 = JOIN(tmp,sal,'emp_no','inner') # ALL
tmp1 = JOIN(tmp,latest_sals,'emp_no','inner') # LATEST

IDs = tmp1['dept_no'].unique()
tmp1 = tmp1[['dept_no','dept_name','salary']]

dfs = []
data = tmp1; ID = 'dept_no'
for i in range (0,len(IDs)):
    tmp= data[data[ID] == IDs[i]]
    tmp['min'] = min(tmp['salary'])
    tmp['max'] = max(tmp['salary'])
    tmp['ave'] = round(tmp['salary'].mean(),0)
    tmp1 = tmp.iloc[[0]]
    dfs.append(tmp1)
tmp = pd.concat(dfs)
out = tmp.sort_values(by=['dept_no'], ascending=True).reset_index()
del out['index']
out

dept_no           dept_name  salary    min     max      ave
0    d001           Marketing   99651  40000  143644  77960.0
1    d002             Finance   95035  40000  134662  76187.0
2    d003     Human Resources   94692  40000  123674  61560.0
3    d004          Production   43699  40000  132552  65725.0
4    d005         Development   88958  40000  144434  65761.0
5    d006  Quality Management   94443  40000  122376  63398.0
6    d007               Sales   72527  40000  158220  86965.0
7    d008            Research   88070  40000  124181  66066.0
8    d009   Customer Service;   56753  40000  144866  64905.0

[Back to contents](#cont)

**<a id='CTE'>Common Table Expressions</a>**

**Example 1: Considering the salary contracts signed by female employees in the company,
how many have been signed for a value below the average? Also provide the total number of contracts signed by women.**

In [50]:
# START WITH AVERAGE SALARY
ave_sal = AGG(sal,'salary','ave'); print(ave_sal)
tmp1 = JOIN(emp,sal,'emp_no','inner')
tmp2 = FILTER(tmp1,'gender','F','=='); #print(tmp2)
ave =  GROUP_COUNT(tmp2,'emp_no');
tmp3 = FILTER(tmp2,'salary',ave_sal,'<'); #print(tmp3)
bel_ave =  GROUP_COUNT(tmp3,'emp_no')
print('Total female salaries = %d, number below average = %d' %(len(ave),len(bel_ave)))

63761.2043294429
Total female salaries = 40832, number below average = 31854


**Exmple 2: Create a temporary table containing the employee number and the lowest salary for each male employee in the company and retrieve all the information stored about employees whose employee numbers are less than 10007.**

In [64]:
tmp1 = JOIN(emp,sal,'emp_no','inner')
tmp2 = FILTER(tmp1,'gender','M','==')
tmp3 = GROUP_AGG(tmp2,'emp_no','salary','min')
tmp3['emp_no'] = tmp3.index
tmp = FILTER(tmp3,'emp_no',10007,'<='); tmp

salary  emp_no
emp_no                
10001    60117   10001
10003    40006   10003
10004    40054   10004
10005    78228   10005

[Back to contents](#cont)

**Example 3: Execute the following two queries**<br>
1. Create a temporary table based on the data in the salaries table. It should contain the following five fields for all employees:<br>

    (a) Employee number<br>

    (b) Salary<br>

    (c) A field named containing the salary rounded to the nearest cent, which should be:<br>

        • Multiplied by 6.5 if the contract start date was between January 1, 1970, and December 31, 1989, inclusive
    
        • Multiplied by 2.8 if the contract start date was between January 1, 1990, and December 31, 1999, inclusive
    
        • Multiplied by 3 for the rest of the contracts

    (d) The contract start date<br>

    (e) The contract end date<br>
    

2. Select all the data from the temporary table just created.<br>

In [81]:
tmp = JOIN(emp,sal,'emp_no','inner'); 
#print(tmp.columns)
tmp = tmp[['emp_no','salary', 'from_date', 'to_date']]
def sal_adj(date1,date2,factor):
    tmp.loc[(tmp['from_date'] >= date1) & (tmp['from_date'] <= date2), 'Adj. salary'] = factor*tmp['salary']
sal_adj("1970-01-01","1989-12-31",6.5)
sal_adj("1990-01-01","1999-12-31",2.8)    
tmp.loc[(tmp['from_date'] < "1970-01-01") | (tmp['from_date'] > "1999-12-31"), 'Adj. salary'] = 3*tmp['salary']
tmp

emp_no  salary   from_date      to_date  Adj. salary
0        10001   60117  1986-06-26   1987-06-26     390760.5
1        10001   62102  1987-06-26   1988-06-25     403663.0
2        10001   66074  1988-06-25   1989-06-25     429481.0
3        10001   66596  1989-06-25   1990-06-25     432874.0
4        10001   66961  1990-06-25   1991-06-25     187490.8
...        ...     ...         ...          ...          ...
967325  201771   56736  1998-04-02   1999-04-02     158860.8
967326  201771   56577  1999-04-02   2000-04-01     158415.6
967327  201771   60056  2000-04-01   2001-04-02     180168.0
967328  201771   62899  2001-04-02   2002-04-01     188697.0
967329  201771   66792  2002-04-01  9999-01-01;     200376.0

[967330 rows x 5 columns]